### PR TITLE
Add UnitActionModifier for Stockpile Cost

### PR DESCRIPTION
**Pull request by itanasi**
Add Stockpile Resource Costs to any Action!
image
Also flip the signs for Action cost to be more consistent. Positive value for Movement means it will cost X Movement, similarly Positive Food will cost (remove) that amount of Food

### CODE REVIEW

File: core/src/com/unciv/models/stats/Stats.kt [hunk 1]
```diff
@@ -186,9 +186,9 @@ open class Stats(
     }
 
     /** Return a string of just +/- value and Stat symbol*/
-    fun toStringOnlyIcons(): String {
+    fun toStringOnlyIcons(plusSign: Boolean = true): String {
         return this.joinToString {
```
Comment (yairm210) [inline core/src/com/unciv/models/stats/Stats.kt:190]: Rename to addPlusSign

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt [hunk 2]
```diff
@@ -140,7 +159,14 @@ object UnitActionModifiers {
...
             for (conditional in actionUnique.conditionals.filter { it.type == UniqueType.UnitActionStatsCost })
                 statCost.add(conditional.stats)
-            effects += (statCost * -1).toStringOnlyIcons()
+            effects += statCost.toStringOnlyIcons(false)
+        }
+
+        if (actionUnique.conditionals.any { it.type == UniqueType.UnitActionStockpileCost }) {
+            var stockpileString = ""
+            for (conditionals in actionUnique.conditionals.filter { it.type == UniqueType.UnitActionStockpileCost })
+                stockpileString += " ${conditionals.params[0].toInt()} [${conditionals.params[1]}]"
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt:168]: Will not translate well, replace with curly brackets

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt [hunk 3]
```diff
@@ -140,7 +159,14 @@ object UnitActionModifiers {
...
+            effects += stockpileString.drop(1) // drop leading space
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt:169]: RemovePrefix, for clarity

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt [hunk 4]
```diff
@@ -63,13 +63,27 @@ object UnitActionModifiers {
         return true
     }
 
+    private fun canSpendStockpileCost(unit: MapUnit, actionUnique: Unique): Boolean {
+        for (conditional in actionUnique.conditionals.filter { it.type == UniqueType.UnitActionStockpileCost }) {
+            val amount = conditional.params[0].toInt()
+            val resourceName = conditional.params[1]
+            if (unit.civ.getCivResourcesByName()[resourceName]!! < amount) {
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt:70]: A malformed resourceName will cause crash, I think we have a safe version? If not then do like 'val amount = ... ?: return false'

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt [hunk 5]
```diff
@@ -97,6 +111,11 @@ object UnitActionModifiers {
                         } else unit.getClosestCity()?.addStat(stat, -value.toInt())
                     }
                 }
+                UniqueType.UnitActionStockpileCost -> {
+                    val amount = conditional.params[0].toInt()
+                    val resourceName = conditional.params[1]
+                    unit.civ.resourceStockpiles.add(resourceName, -amount)
```
Comment (yairm210) [inline core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt:117]: Check for valid resource?

File: core/src/com/unciv/models/stats/Stats.kt (hunk 1 above)
Comment (itanasi) [reply core/src/com/unciv/models/stats/Stats.kt:190]: Done

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt (hunk 2 above)
Comment (itanasi) [reply core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt]: Check again?

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt (hunk 3 above)
Comment (itanasi) [reply core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt]: Done

File: core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt (hunk 4 above)
Comment (itanasi) [reply core/src/com/unciv/ui/screens/worldscreen/unit/actions/UnitActionModifiers.kt]: Found the safety version

**Review by yairm210 — APPROVED** [review]

**Pull request merged**